installer/installer_centos.sh

```diff
@@ -1,3 +1,141 @@
 #!/usr/bin/env bash
 
+log(){
+    echo "$1" | tee -a installer_centos.log
+}
 
+# this only runs on centos 
+if ! grep -i centos /etc/os-release;then
+    log "this now only supports centos"
+    exit 1
+fi
+
+# change the yum source
+# this only works for centos6 and centos7
+# the original source.list will be renamed with source.list.bak
+sudo_cmd=$(command -v sudo)
+if grep 'VERSION="7 (Core)' /etc/os-release; then
+    ${sudo_cmd} cp /etc/yum/CentOS-Base.repo /etc/yum/CentOS-Base.repo.bak
+    ${sudo_cmd} sed -e 's|^mirrorlist=|#mirrorlist=|g' \
+    -e 's|^#baseurl=http://mirror.centos.org/centos|baseurl=https://mirrors.tuna.tsinghua.edu.cn/centos|g' \
+    -i.bak \
+    /etc/yum.repos.d/CentOS-*.repo | ${sudo_cmd} tee /etc/yum/CentOS-Base.repo || exit 1
+    ${sudo_cmd} yum update || exit 1
+elif grep 'VERSION="8"' /etc/os-release;then
+    ${sudo_cmd} cp /etc/yum.repos.d/CentOS-Linux-BaseOS.repo /etc/yum.repos.d/CentOS-Linux-BaseOS.repo.bak
+    ${sudo_cmd} sed -e 's|^mirrorlist=|#mirrorlist=|g' \
+    -e 's|^#baseurl=http://mirror.centos.org/\$releasever|baseurl=https://mirrors.tuna.tsinghua.edu.cn/centos|g' \
+    -i.bak \
+    /etc/yum.repos.d/CentOS-*.repo | ${sudo_cmd} tee /etc/yum/CentOS-Base.repo || exit 1
+    ${sudo_cmd} yum update || exit 1
+else
+    log "only support centos7 and centos8,change the source by yourself"
+fi
+
+#install sudo
+if !  ${sudo_cmd} --version;then
+    ${sudo_cmd} yum install -y sudo
+fi
+
+#basic tools
+${sudo_cmd} yum install -y curl wget tar ripgrep || exit 1
+
+
+# this is for add-yum-repository
+sudo yum install -y yum-utils || exit 1
+if grep 'VERSION="7 (Core)' /etc/os-release; then
+    sudo yum install -y epel-release || exit 1
+fi
+# install fish 3
+if ! fish --version;then
+
+    url="https://download.opensuse.org/repositories/shells:fish:release:3/CentOS_"
+    if grep 'VERSION="8"' /etc/os-release; then
+        url="${url}8/shells:fish:release:3.repo"
+     #centos6
+    elif grep 'VERSION="7 (Core)' /etc/os-release; then
+        url="${url}7/shells:fish:release:3.repo"
+    else 
+        url=""
+    fi
+    if [ -z "$url" ]; then
+        echo "only support CentOS 8 and 7"
+        exit 1
```

Comment:
This should not use `exit 1` to terminate the script, because `fish` is not necessary for all users. You should just add a log that dicates the fish is not installed, and could be installed mannually.